### ISSUE TITLE
Failed to create path

### ISSUE THREAD

**Issue by varenius** [jive-vlbi/etransfer#18]
We have an interesting situation where I cannot send data from Onsala to Tsukuba using etc/etd. I really want to understand why. Source (etc) and destination (etd) are both version 1.0.1. Below are logs from both etc and etd sides. In the log output, etc IP has been replaced with SOUR and etd ip has been replaced by DEST.

ETC:
```
oper@gyller:~$ etc '/mnt/tsukuba/b21096_oe*' tcp://DEST#2642:/export/vlbi/data/b21096oe/ -m 5 --resume
2021-09-10 12:08:10.93: [etdc::ETDServer::ETDServer(etdc::etd_state&)] ETDServer starting, my uuid=AisriHyVXGOMDQVki
2021-09-10 12:08:10.93: [etdc::etdc_fdptr mk_client(const T&, const etdc::detail::client_settings&) [with T = etdc::protocol_type; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] mk_client/attempt #1/3 trying to connect to tcp:DEST:2642
2021-09-10 12:08:11.10: [virtual etdc::protocolversion_type etdc::ETDProxy::protocolVersion() const] ETDProxy::protocolVersion/sending message 'protocol-version
'
2021-09-10 12:08:11.28: [int main(int, const char* const*)] This client supports protocol version 1
2021-09-10 12:08:11.28: [int main(int, const char* const*)] Server protocol version: 1
2021-09-10 12:08:11.28: [int main(int, const char* const*)] Server protocol version: 1
2021-09-10 12:08:11.32: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] ETDProxy::dataChannelAddr/sending message 'data-channel-addr-ext
'
2021-09-10 12:08:11.48: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] dataChannelAddr/reply from server: 'OK <tcp/0.0.0.0:2644/mss=1500,max-bw=-1>'
2021-09-10 12:08:11.49: [etdc::sockname_type etdc::decode_data_addr(const string&)] decode_data_addr: 1='tcp' 2='0.0.0.0' 9='2644' 11='mss=1500,max-bw=-1'
2021-09-10 12:08:11.49: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] dataChannelAddr/reply from server: 'OK'
2021-09-10 12:08:11.49: [int main(int, const char* const*)] PUSH Resume /mnt/tsukuba/b21096_oe_096-1830a_0 -> /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0
2021-09-10 12:08:11.49: [void signal_thread(const signallist_type&, pthread_t, etdc::etd_state&, std::vector<std::shared_ptr<etdc::ETDServerInterface> >&, unique_result (&)[2]) [with int KillSignal = 10; signallist_type = std::vector<int>; pthread_t = long unsigned int; unique_result = std::unique_ptr<std::tuple<etdc::uuid_type, long int> >]] sigwaiterthread: enter wait phase
2021-09-10 12:08:11.49: [virtual etdc::result_type etdc::ETDProxy::requestFileWrite(const string&, etdc::openmode_type)] ETDProxy::requestFileWrite/sending message 'write-file-Resume /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0
' sz=67
2021-09-10 12:08:11.66: [int main(int, const char* const*)] Got exception: assertion error: src/etdc_etdserver.cc:935 [status_s=="OK"] requestFileWrite(/export/vlbi/data/b21096oe/b21096_oe_096-1830a_0) failed - assertion error: /usr/local/src/etransfer/etransfer-1.0.1/src/etdc_fd.h:1288 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096oe' -
2021-09-10 12:08:11.66: [int main(int, const char* const*)] Retry #2 (#2 for this file), go to sleep for 10s
2021-09-10 12:08:21.66: [int main(int, const char* const*)] PUSH Resume /mnt/tsukuba/b21096_oe_096-1830a_0 -> /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0
2021-09-10 12:08:21.66: [virtual etdc::result_type etdc::ETDProxy::requestFileWrite(const string&, etdc::openmode_type)] ETDProxy::requestFileWrite/sending message 'write-file-Resume /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0
' sz=67
2021-09-10 12:08:21.84: [int main(int, const char* const*)] Got exception: assertion error: src/etdc_etdserver.cc:935 [status_s=="OK"] requestFileWrite(/export/vlbi/data/b21096oe/b21096_oe_096-1830a_0) failed - assertion error: /usr/local/src/etransfer/etransfer-1.0.1/src/etdc_fd.h:1288 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096oe' -
2021-09-10 12:08:21.84: [etdc::etd_state::~etd_state()] ~etd_state/need to wait for 0 threads 
```

ETD:
```
2021-09-10 12:08:11.18: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] Incoming COMMAND from <tcp/SOUR:57388> [local <tcp/DEST:2642>]
2021-09-10 12:08:11.18: [<lambda(etdc::etdc_fdptr, const string&)>] client/TCP rcvbuf = 67108864 sndbuf = 67108864
2021-09-10 12:08:11.18: [etdc::ETDServer::ETDServer(etdc::etd_state&)] ETDServer starting, my uuid=6sB6tJMAgwdNRYzRotLr
2021-09-10 12:08:11.18: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-10 12:08:11.18: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=17 => nTotal=17
2021-09-10 12:08:11.18: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'protocol-version'
2021-09-10 12:08:11.19: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK 1'
2021-09-10 12:08:11.19: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-10 12:08:11.40: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=22 => nTotal=22
2021-09-10 12:08:11.40: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'data-channel-addr-ext'
2021-09-10 12:08:11.40: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK <tcp/0.0.0.0:2644/mss=1500,max-bw=-1>'
2021-09-10 12:08:11.40: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-09-10 12:08:11.40: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-10 12:08:11.57: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=67 => nTotal=67
2021-09-10 12:08:11.57: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'write-file-Resume /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0'
2021-09-10 12:08:11.57: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/npath='/export/vlbi/data/b21096oe/b21096_oe_096-1830a_0'
2021-09-10 12:08:11.57: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/O_CREAT is set, dir='/export/vlbi/data/b21096oe/'
2021-09-10 12:08:11.57: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export'
2021-09-10 12:08:11.57: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi'
2021-09-10 12:08:11.58: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data'
2021-09-10 12:08:11.58: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data/b21096oe'
2021-09-10 12:08:11.58: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'ERR assertion error: /usr/local/src/etransfer/etransfer-1.0.1/src/etdc_fd.h:1288 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096oe' - '
2021-09-10 12:08:11.58: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-10 12:08:21.75: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=67 => nTotal=67
2021-09-10 12:08:21.75: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'write-file-Resume /export/vlbi/data/b21096oe/b21096_oe_096-1830a_0'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/npath='/export/vlbi/data/b21096oe/b21096_oe_096-1830a_0'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/O_CREAT is set, dir='/export/vlbi/data/b21096oe/'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data'
2021-09-10 12:08:21.75: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data/b21096oe'
2021-09-10 12:08:21.75: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'ERR assertion error: /usr/local/src/etransfer/etransfer-1.0.1/src/etdc_fd.h:1288 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096oe' - '
2021-09-10 12:08:21.75: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-10 12:08:21.92: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=0 => nTotal=0
2021-09-10 12:08:21.92: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] command server thread got exception: assertion error: src/etdc_etdserver.cc:1291 [n>0] Failed to read data from remote end
2021-09-10 12:08:21.92: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] command server thread terminated
```

Naïvely, I interpret the logs to say that the receiving etd process does not have write permission in the requested target folder, OR the target folder does not exist (and etd is not allowed to create the folder). However, according to Tsukuba staff (etd side), the path does exist and is writeable; `ls -l` output:
`drwxrwxrwx. 2 vlbi vlbi         6 Sep 14 15:32 b21096oe`. 

Do you have any idea what could be causing this? It is a show-stopper for a lot of transfer to Tsukuba.

**Comment by haavee** [comment]
Weird! Thanks for the detailed log, is helpful.

Two observations:
- it is important to know exactly `etd` was started and by whom (e.g. if daemonized could have been run as different user)
- the strange thing is that the _actual_ error message seems to be missing; there's an opening `'` followed by nothing? I'd be very interested in the string error message ... it might shed some light on why the call #FAILSs

```
2021-09-10 12:08:21.75: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'ERR assertion error: 
/usr/local/src/etransfer/etransfer-1.0.1/src/etdc_fd.h:1288 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to 
create path '/export/vlbi/data/b21096oe' - '
```

**Comment by haavee** [comment]
Could you ask Ishioka to apply the attached patch:
```
> cd /path/to/etransfer
> patch < /path/to/downloaded/more-info.txt
> make
```
and see if that turns up more useful information and/or fixes the issue
[more-info.txt](https://github.com/jive-vlbi/etransfer/files/7185552/more-info.txt)
?

**Comment by varenius** [comment]
Ishioka (receiver) has applied patch. It seems to both turn up more useful info AND fix the issue :). I'll try to clarify what I have learned below:

I wanted to run the command `oper@gyller:~$ etc '/mnt/tsukuba/b21096_oe*' tcp://DEST#2642:/export/vlbi/data/b21096oe/ -m 5 --resume` but ran into issues (see top of this ticket). 

After the patch was applied, I wanted to try the same command again, but by mistake I had a different target directory (which should not exist); I ran `oper@gyller:~$ etc '/mnt/tsukuba/b21096_oe*' tcp://DEST#2642:/export/vlbi/data/b21096_ow/ -m 5 --resume`. With this command, the server log reported (correctly) that the sought directory wasn't found, but this time with a little more information in the form of `[errno=13]`:
```
2021-09-27 11:40:22.42: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] Incoming COMMAND from <tcp/130.242.10.13:33638> [local <tcp/210.146.79.7:2642>]
2021-09-27 11:40:22.42: [<lambda(etdc::etdc_fdptr, const string&)>] client/TCP rcvbuf = 67108864 sndbuf = 67108864
2021-09-27 11:40:22.42: [etdc::ETDServer::ETDServer(etdc::etd_state&)] ETDServer starting, my uuid=aqWEx2YqOcDrl2O
2021-09-27 11:40:22.42: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-27 11:40:22.42: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=17 => nTotal=17
2021-09-27 11:40:22.42: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'protocol-version'
2021-09-27 11:40:22.43: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK 1'
2021-09-27 11:40:22.43: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-27 11:40:22.67: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=22 => nTotal=22
2021-09-27 11:40:22.67: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'data-channel-addr-ext'
2021-09-27 11:40:22.67: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK <tcp/0.0.0.0:2644/mss=1500,max-bw=-1>'
2021-09-27 11:40:22.67: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-09-27 11:40:22.67: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-27 11:40:22.86: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=68 => nTotal=68
2021-09-27 11:40:22.86: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'write-file-Resume /export/vlbi/data/b21096_ow/b21096_oe_096-1830a_0'
2021-09-27 11:40:22.86: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/npath='/export/vlbi/data/b21096_ow/b21096_oe_096-1830a_0'
2021-09-27 11:40:22.86: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/O_CREAT is set, dir='/export/vlbi/data/b21096_ow/'
2021-09-27 11:40:22.86: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export'
2021-09-27 11:40:22.86: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi'
2021-09-27 11:40:22.87: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data'
2021-09-27 11:40:22.87: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data/b21096_ow'
2021-09-27 11:40:22.87: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'ERR assertion error: /usr/local/src/etransfer/etransfer-1.0.1_more-info/src/etdc_fd.h:1289 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096_ow' -  [errno=13]'
2021-09-27 11:40:22.87: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-27 11:40:33.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=68 => nTotal=68
2021-09-27 11:40:33.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'write-file-Resume /export/vlbi/data/b21096_ow/b21096_oe_096-1830a_0'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/npath='/export/vlbi/data/b21096_ow/b21096_oe_096-1830a_0'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/O_CREAT is set, dir='/export/vlbi/data/b21096_ow/'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data'
2021-09-27 11:40:33.05: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {int}; std::string = std::basic_string<char>]] open_file/path_so_far='/export/vlbi/data/b21096_ow'
2021-09-27 11:40:33.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'ERR assertion error: /usr/local/src/etransfer/etransfer-1.0.1_more-info/src/etdc_fd.h:1289 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21096_ow' -  [errno=13]'
2021-09-27 11:40:33.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-09-27 11:40:33.24: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=0 => nTotal=0
2021-09-27 11:40:33.24: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] command server thread got exception: assertion error: src/etdc_etdserver.cc:1291 [n>0] Failed to read data from remote end
2021-09-27 11:40:33.24: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] command server thread terminated
```
Realising my error, I then tried with the correct command that I wanted to work all along, and now the transfers started without problems!

Conclusions:
- With the patch, we get more info on the server side.
- With the patch, my original transfer works!

**Comment by varenius** [comment]
I spoke too soon :(. I transferred some data yesterday without issues, so I thought the "more-info" patch fixed the problem. Then, later yestertday evening, the issue suddenly came back. Below is the log output from my side (etc).

FIRST PART: From a successful transfer that suddenly stops with "Connection timed out"
```
[..]
2021-09-28 18:48:02.93: PUSH Resume /mnt/tsukuba/b21102_ow_102-1902b_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1902b_0
finished; succesfully transferred 1.5 GiB (1,469,604,128 bytes) in 28.51 s [49.16 MBps]
2021-09-28 18:48:32.34: PUSH Resume /mnt/tsukuba/b21102_ow_102-1902b_1 -> /export/vlbi/data/b21102ow/b21102_ow_102-1902b_1
finished; succesfully transferred 1.5 GiB (1,469,612,352 bytes) in 23.39 s [59.92 MBps]
2021-09-28 18:48:56.62: PUSH Resume /mnt/tsukuba/b21102_ow_102-1902b_2 -> /export/vlbi/data/b21102ow/b21102_ow_102-1902b_2
finished; succesfully transferred 1.5 GiB (1,469,563,008 bytes) in 21.97 s [63.78 MBps]
2021-09-28 18:49:19.49: PUSH Resume /mnt/tsukuba/b21102_ow_102-1902b_3 -> /export/vlbi/data/b21102ow/b21102_ow_102-1902b_3
Unfinished; succesfully transferred 1.4 GiB (1,409,286,144 bytes) in 1.064 ks [1.26 MBps]
--> Reason: Connection timed out
2021-09-28 23:46:11.35: PUSH Resume /mnt/tsukuba/b21102_ow_102-1902b_3 -> /export/vlbi/data/b21102ow/b21102_ow_102-1902b_3
```

SECOND PART: Trying to restart the transfer with the resume option, I don't get any transfer starting:
```
oper@gyller:~$ etc '/mnt/tsukuba/b21102_ow*' DEST#2642:/export/vlbi/data/b21102ow/ --resume
2021-09-29 05:10:48.09: PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
2021-09-29 05:10:58.26: PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
```

Same again, no luck:
```
oper@gyller:~$ etc '/mnt/tsukuba/b21102_ow*' DEST#2642:/export/vlbi/data/b21102ow/ --resume
2021-09-29 05:11:08.29: PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
2021-09-29 05:11:18.46: PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
```

THIRD PART: Trying again with "-m 5" to get more log output. We see that we again have the "failed to create path" error:
```
oper@gyller:~$ etc '/mnt/tsukuba/b21102_ow*' DEST#2642:/export/vlbi/data/b21102ow/ -m 5 --resume
2021-09-29 05:11:23.78: [etdc::ETDServer::ETDServer(etdc::etd_state&)] ETDServer starting, my uuid=Zx13I9IOjlKLytMeJ9
2021-09-29 05:11:23.78: [etdc::etdc_fdptr mk_client(const T&, const etdc::detail::client_settings&) [with T = etdc::protocol_type; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] mk_client/attempt #1/3 trying to connect to tcp:DEST:2642
2021-09-29 05:11:23.95: [virtual etdc::protocolversion_type etdc::ETDProxy::protocolVersion() const] ETDProxy::protocolVersion/sending message 'protocol-version
'
2021-09-29 05:11:24.12: [int main(int, const char* const*)] This client supports protocol version 1
2021-09-29 05:11:24.12: [int main(int, const char* const*)] Server protocol version: 1
2021-09-29 05:11:24.12: [int main(int, const char* const*)] Server protocol version: 1
2021-09-29 05:11:24.18: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] ETDProxy::dataChannelAddr/sending message 'data-channel-addr-ext
'
2021-09-29 05:11:24.35: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] dataChannelAddr/reply from server: 'OK <tcp/0.0.0.0:2644/mss=1500,max-bw=-1>'
2021-09-29 05:11:24.35: [etdc::sockname_type etdc::decode_data_addr(const string&)] decode_data_addr: 1='tcp' 2='0.0.0.0' 9='2644' 11='mss=1500,max-bw=-1'
2021-09-29 05:11:24.35: [virtual etdc::dataaddrlist_type etdc::ETDProxy::dataChannelAddr() const] dataChannelAddr/reply from server: 'OK'
2021-09-29 05:11:24.35: [int main(int, const char* const*)] PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
2021-09-29 05:11:24.35: [virtual etdc::result_type etdc::ETDProxy::requestFileWrite(const string&, etdc::openmode_type)] ETDProxy::requestFileWrite/sending message 'write-file-Resume /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
' sz=67
2021-09-29 05:11:24.35: [void signal_thread(const signallist_type&, pthread_t, etdc::etd_state&, std::vector<std::shared_ptr<etdc::ETDServerInterface> >&, unique_result (&)[2]) [with int KillSignal = 10; signallist_type = std::vector<int>; pthread_t = long unsigned int; unique_result = std::unique_ptr<std::tuple<etdc::uuid_type, long int> >]] sigwaiterthread: enter wait phase
2021-09-29 05:11:24.52: [int main(int, const char* const*)] Got exception: assertion error: src/etdc_etdserver.cc:935 [status_s=="OK"] requestFileWrite(/export/vlbi/data/b21102ow/b21102_ow_102-1830a_0) failed - assertion error: /usr/local/src/etransfer/etransfer-1.0.1_more-info/src/etdc_fd.h:1289 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21102ow' -  [errno=13]
2021-09-29 05:11:24.52: [int main(int, const char* const*)] Retry #2 (#2 for this file), go to sleep for 10s
2021-09-29 05:11:34.52: [int main(int, const char* const*)] PUSH Resume /mnt/tsukuba/b21102_ow_102-1830a_0 -> /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
2021-09-29 05:11:34.52: [virtual etdc::result_type etdc::ETDProxy::requestFileWrite(const string&, etdc::openmode_type)] ETDProxy::requestFileWrite/sending message 'write-file-Resume /export/vlbi/data/b21102ow/b21102_ow_102-1830a_0
' sz=67
2021-09-29 05:11:34.69: [int main(int, const char* const*)] Got exception: assertion error: src/etdc_etdserver.cc:935 [status_s=="OK"] requestFileWrite(/export/vlbi/data/b21102ow/b21102_ow_102-1830a_0) failed - assertion error: /usr/local/src/etransfer/etransfer-1.0.1_more-info/src/etdc_fd.h:1289 [::mkdir(path_so_far.c_str(), 0755)==0 || errno==EEXIST] Failed to create path '/export/vlbi/data/b21102ow' -  [errno=13]
2021-09-29 05:11:34.69: [etdc::etd_state::~etd_state()] ~etd_state/need to wait for 0 threads
```
I've asked for the server-side logs via email.

**Comment by varenius** [comment]
Interesting progress on this topic happened this week. First from Ishioka colleagues who did a wonderful job of pinning down the root cause of this bug:
> Regarding to the data transfer, we realized the problem that is occurring under the specific situation.
> This seems to be a reason for the fail to transfer.
>
> If it is the first time to access the nfs volume after it was mounted,
> and if the designated directory was already exist,
> the server send the error message 'errno=13(EACCES)' when the etc/etd program runs 'mkdir'.
> It seems to cause 'Failed to create path'.
>
> Once we access to the designated directory with 'ls' command, the message 'errno=17(EEXIST)' is returned when 'mkdir' runs and the program runs normally.
>
>
> When the 'NFS server not responding...' is occurred, it is the same status as the nfs server is remounted, so the transfer failed in the middle.
>
>
> To solve this problem, we add the 'stat' function to read the attribute information of the directory before 'mkdir' in the program.
> We checked that the problem seems to be fixed in the test transfer with our local network.
>
> So, we would like to do test transfer with your server again.
> If it is okay with you, could you try to transfer the data?

I tested transferring and resuming many times, without any errors! So it seems they indeed found the root cause. @haavee replied that 

> This is very good news, excellent debugging skills displayed!
> Of course it would be very simple to add a stat(2) call in the function that checks/creates the path.
>
> I'll do some digging first since it does sound a bit like it's a dodgy issue.
> I found this thread:
>
>     https://bugs.python.org/issue14702
>
> which seems to run into exactly the same problem we're facing and people have done strace() to see the #FAIL happening.
>
>  From that discussion it seems that the easiest way is a stat(2) on the full path, before recursing into creating the directories that don't exist (yet).

So it seems this is indeed fixable in etd. Great, now @haavee just need to find the time to implement the fix in this official repo. Even though it may take some time before we get there, I'm very happy that we seem to have found the root cause for this!

**Comment by haavee** [comment]
Please verify that this branch https://github.com/jive-vlbi/etransfer/tree/issue-18 fixes the issue, then it can be merged into master.

**Comment by haavee** [comment]
Ah. The _actual_ fix was still uncommitted sitting on my disk as local mod. Should now be committed and pushed to the `issue-18` branch, sorrie!

**Comment by varenius** [comment]
After applying the proposed patch, i.e. compiling and running the issue-18 branch, the problem appears to be gone! I did some extensive testing this morning with our colleagues in Japan without finding any issues. We have also done local testing at OSO and GSI. I hence think we can merge this patch into main branch. I also think we should make another proper release, since this bug may affect other correlators with NFS mounted storage.

**Comment by haavee** [comment]
This issue is closed by the v1.1 release (see https://github.com/jive-vlbi/etransfer/commit/fc1472c12ba374a6e57ae46f5d48f177341ae3cc)
